### PR TITLE
PS-6112: Inconsistent Binlog_snapshot_gtid when mysqldump used with --single-transaction (8.0)

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PS-6112

When mysqldump is started with --single-transaction option it starts repeatable read transaction, so dump is expected to be consistent. Parallel transactions do not affect data visible by mysqldump transaction, however they modify gtid_executed variable. As gtid_executed variable is not part of the transaction, mysqldump captured its updated value.

The fix is to use 'Binlog_snapshot_gtid_executed' variable introduced by PS-4464. If the variable is not supported by server, behavior falls back to upstream's mysqldump behavior.

Additionally there was a bug when starting mysqldump with --master-data  --lock-for-backup and --single-transaction. UNLOCK TABLES query was issued just after collecting binlog coordinates instead of keeping tables blocked up to the end of backup.

Two options were added to mysqldump: start-sql-file and finish-sql-file. This allows better integration of mysqldump with MTR tests as debug sync facility as server sync points are activated per thread.

This is not GCA commit because it needs changes done in PS-4464

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/520/
Failed MTR tests not related to the change.

File: client/mysqldump.cc [hunk 1]
```diff
@@ -1252,6 +1269,30 @@ static int fetch_db_collation(const char *db_name, char *db_cl_name,
...
 }
 
+/*
+  Check if server supports binlog_snapshot_gtid_executed.
+  Returns 1 supported, 0 if not.
+ */
+static int consistent_gtid_executed_supported(MYSQL *mysql_con) {
+  MYSQL_RES *res;
+  MYSQL_ROW row;
+  int found = 0;
```
Comment (percona-ysorokin) [inline client/mysqldump.cc]:
```suggestion
  bool found = 0;
```

**Review by percona-ysorokin — APPROVED** [review]
LGTM with minor comment (can be ignored)

File: client/mysqldump.cc (hunk 1 above)
Comment (kamil-holubicki) [reply client/mysqldump.cc]: done

**Pull request merged**